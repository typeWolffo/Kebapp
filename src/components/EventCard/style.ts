import tw from 'tailwind-styled-components'

export const StyledEvent = tw.div`
    w-9/12
    p-5
    border border-solid ${({ $isDraft }: { $isDraft: boolean }) => ($isDraft ? 'border-secondary' : 'border-primary')}
    rounded-md mb-5
`
export const StyledHeader = tw.div`
    w-full
    text-xl
    flex
    justify-between
`

export const StyledIcon = tw.div`
    w-8
    h-8
`

export const StyledContent = tw.div`
    ${({ $isActive }: { $isActive: boolean }) => ($isActive ? 'h-fit' : 'h-0')}
    overflow-hidden
`

export const StyledParticipantsWrapper = tw.div`first-letter:
    mt-4
    px-1
    flex
`

export const StyledParticipant = tw.div`first-letter:
    border border-solid border-primary
    w-8
    h-8
    text-lg
    rounded-[100%]
    flex items-center justify-center
    -mx-1
`

export const StyledButton = tw.button`
    btn btn-outline btn-accent
    mt-4
    w-full
`
