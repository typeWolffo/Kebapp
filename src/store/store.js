import { configureStore } from "@reduxjs/toolkit";
import authReducer from "../slices/auth";
import messageReducer from "../slices/message";
import eventReducer from "../slices/event";

const reducer = {
  auth: authReducer,
  event: eventReducer,
  message: messageReducer,
};

const store = configureStore({
  devTools: true,
  reducer,
});

export default store;
