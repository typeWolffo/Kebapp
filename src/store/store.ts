import { configureStore } from '@reduxjs/toolkit'
import authReducer from '../slices/auth'
import messageReducer from '../slices/message'
import eventReducer from '../slices/event'
import modalReducer from '../slices/modal'
import getAllEventsReducer from '../slices/getAllEvents'
import getSingleEventReducer from '../slices/getSingleEvent'
import { useDispatch } from 'react-redux'

const reducer = {
  auth: authReducer,
  event: eventReducer,
  getAllEvents: getAllEventsReducer,
  getSingleEvent: getSingleEventReducer,
  message: messageReducer,
  modal: modalReducer,
}

export const store = configureStore({
  devTools: true,
  reducer,
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export const useAppDispatch = () => useDispatch<AppDispatch>()
