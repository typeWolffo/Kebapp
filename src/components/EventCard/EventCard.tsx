import { ChevronDown } from '@styled-icons/boxicons-regular/ChevronDown'
import isEmpty from 'lodash.isempty'
import { memo, useCallback, useState } from 'react'
import { useDispatch } from 'react-redux'
import { useNavigate } from 'react-router-dom'

import { useJoinEvent } from '../../hooks/eventHooks'
import { FormModes, setFormMode } from '../../slices/form'
import { EventDataType } from '../../types/EventType'
import { UserType } from '../../types/UserType'
import { StyledButton, StyledContent, StyledEvent, StyledHeader, StyledIcon, StyledParticipant, StyledParticipantsWrapper } from './style'



type Props = {
  event: EventDataType
  currentUserStatus: string
  currentUser: UserType
}

const getKebsDate = (event: Date) => {
  const date = new Date(event)

  return {
    date: `${date.getDate()}.${date.getMonth() + 1}.${date.getFullYear()}`,
    day: String(date.getDay()),
    hour: String(date.getHours()),
    minute: String(date.getMinutes()),
  }
}

const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']

const weekday = (date: Date) => days[new Date(date).getDay()]

const createLetterAvatar = (name: string) => {
  return <span className="transform -translate-y-[3%]">{name.charAt(0)}</span> // magic number for optical alignment
}

function EventCard(props: Props) {
  const { event, currentUser, currentUserStatus } = props
  const [isActive, setIsActive] = useState(false)
  const { mutate: joinToEvent } = useJoinEvent()
  const navigate = useNavigate()
  const dispatch = useDispatch()

  const handelJoinClick = useCallback(() => joinToEvent(event.id as number), [event.id])
  const handleEditClick = () => {
    dispatch(setFormMode(FormModes.EDIT))
    navigate(`/${event.id}/edit`)
  }

  return (
    <StyledEvent>
      {event && currentUserStatus === 'success' && (
        <>
          <StyledHeader onClick={() => setIsActive(!isActive)}>
            {event.location}
            <StyledIcon>
              <ChevronDown className={isActive ? 'tranfororm  rotate-180' : ''} />
            </StyledIcon>
          </StyledHeader>

          <StyledContent $isActive={isActive}>
            <div className="py-5">
              <span className="mr-2">{`${getKebsDate(event.start_at as Date).hour}:${getKebsDate(event.start_at as Date).minute.padStart(2, '0')}`}</span>
              <span className="mr-2">{`${getKebsDate(event.start_at as Date).date}`}</span>
              <span>{weekday(event.start_at as Date)}</span>
            </div>
            <StyledParticipantsWrapper>
              <StyledParticipant>{createLetterAvatar(event.author?.name as string)}</StyledParticipant>

              {!isEmpty(event.members) && event.members?.map((member: { user: UserType }) => <StyledParticipant key={member.user.id}>{createLetterAvatar(member.user.name)}</StyledParticipant>)}
            </StyledParticipantsWrapper>

            {event.author?.id !== currentUser.id && (
              <StyledButton type="button" onClick={handelJoinClick}>
                Join
              </StyledButton>
            )}
            {event.author?.id === currentUser.id && (
              <StyledButton type="button" onClick={handleEditClick}>
                Edit
              </StyledButton>
            )}
          </StyledContent>
        </>
      )}
    </StyledEvent>
  )
}

export default memo(EventCard)
