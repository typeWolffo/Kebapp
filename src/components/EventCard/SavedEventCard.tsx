import { ChevronDown } from '@styled-icons/boxicons-regular'
import { useState } from 'react'
import { ManageEventType } from '../../types/EventType'
import { StyledContent, StyledEvent, StyledHeader, StyledIcon } from './style'

type Props = {
  event: ManageEventType
  isDraft: boolean
}

const getKebsDate = (event: Date) => {
  const date = new Date(event)

  return {
    date: `${date.getDate()}.${date.getMonth()}.${date.getFullYear()}`,
    day: String(date.getDay()),
    hour: String(date.getHours()),
    minute: String(date.getMinutes()),
  }
}

const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday']

const weekday = (date: Date) => days[new Date(date).getDay()]

function SavedEventCard(props: Props) {
  const { event, isDraft } = props
  const [isActive, setIsActive] = useState(false)

  return (
    <StyledEvent $isDraft={isDraft}>
      {event && (
        <>
          <StyledHeader onClick={() => setIsActive(!isActive)}>
            {event.location}
            <StyledIcon>
              <ChevronDown className={isActive ? 'tranfororm  rotate-180' : ''} />
            </StyledIcon>
          </StyledHeader>

          <StyledContent $isActive={isActive}>
            <div className="py-5">
              <span className="mr-2">{`${getKebsDate(event.start_at as Date).hour}:${getKebsDate(event.start_at as Date).minute.padStart(2, '0')}`}</span>
              <span>{weekday(event.start_at as Date)}</span>
            </div>
          </StyledContent>
        </>
      )}
    </StyledEvent>
  )
}

export default SavedEventCard
