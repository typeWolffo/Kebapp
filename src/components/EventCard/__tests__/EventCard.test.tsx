import { render, screen } from '@testing-library/react'
import { createMemoryHistory } from 'history'
import { QueryClient, QueryClientProvider } from 'react-query'
import { Provider } from 'react-redux'
import { Router } from 'react-router-dom'

import { store } from '../../../store/store'
import EventCard from '../EventCard'

const mockedEvent = {
  id: 8,
  location: 'Alanya Kebab',
  start_at: new Date('2022-06-25 17:18:00'),
  created_at: '2022-06-18T15:19:04.000000Z',
  updated_at: '2022-06-18T15:19:04.000000Z',
  is_active: true,
  is_participating: true,
  author: {
    id: 3,
    name: 'wilk',
    email: 'jw@jw.com',
  },
  members: [
    {
      id: 2,
      event_id: 8,
      user: {
        id: 1,
        name: 'test',
        email: 'test@test.com',
      },
    },
    {
      id: 3,
      event_id: 8,
      user: {
        id: 2,
        name: 'xyz',
        email: 'xyz@xyz.com',
      },
    },
  ],
}

const mockedCurrentUser1 = {
  id: 3,
  name: 'wilk',
  email: 'jw@jw.com',
}
const mockedCurrentUser2 = {
  id: 4,
  name: 'test',
  email: 'testtest@jw.com',
}

describe('EventCard', () => {
  const queryClient = new QueryClient()
  const history = createMemoryHistory()

  it('should render', () => {
    const { container } = render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <QueryClientProvider client={queryClient}>
            <EventCard event={mockedEvent} currentUserStatus={'success'} currentUser={mockedCurrentUser1} />
          </QueryClientProvider>
        </Router>
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })

  it('should render with date', () => {
    render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <QueryClientProvider client={queryClient}>
            <EventCard event={mockedEvent} currentUserStatus={'success'} currentUser={mockedCurrentUser1} />
          </QueryClientProvider>
        </Router>
      </Provider>
    )

    expect(screen.getByText('25.6.2022')).toBeInTheDocument()
  })

  it('should render with weekday', () => {
    render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <QueryClientProvider client={queryClient}>
            <EventCard event={mockedEvent} currentUserStatus={'success'} currentUser={mockedCurrentUser1} />
          </QueryClientProvider>
        </Router>
      </Provider>
    )

    expect(screen.getByText('Saturday')).toBeInTheDocument()
  })

  it('should render with join button', () => {
    render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <QueryClientProvider client={queryClient}>
            <EventCard event={mockedEvent} currentUserStatus={'success'} currentUser={mockedCurrentUser2} />
          </QueryClientProvider>
        </Router>
      </Provider>
    )

    expect(screen.getByRole('button', { name: 'Join' })).toBeInTheDocument()
  })

  it('should render with edit button', () => {
    render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <QueryClientProvider client={queryClient}>
            <EventCard event={mockedEvent} currentUserStatus={'success'} currentUser={mockedCurrentUser1} />
          </QueryClientProvider>
        </Router>
      </Provider>
    )

    expect(screen.getByRole('button', { name: 'Edit' })).toBeInTheDocument()
  })

  it('should render with members', () => {
    render(
      <Provider store={store}>
        <Router location={history.location} navigator={history}>
          <QueryClientProvider client={queryClient}>
            <EventCard event={mockedEvent} currentUserStatus={'success'} currentUser={mockedCurrentUser1} />
          </QueryClientProvider>
        </Router>
      </Provider>
    )

    expect(screen.getByText('w')).toBeInTheDocument()
    expect(screen.getByText('t')).toBeInTheDocument()
    expect(screen.getByText('x')).toBeInTheDocument()
  })
})
