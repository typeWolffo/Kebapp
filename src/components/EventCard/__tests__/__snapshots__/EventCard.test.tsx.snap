// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventCard should render 1`] = `
<div>
  <div
    class="w-9/12 p-5 border border-solid border-primary rounded-md mb-5"
  >
    <div
      class="w-full text-xl flex justify-between"
    >
      Alanya Kebab
      <div
        class="w-8 h-8"
      >
        <svg
          aria-hidden="true"
          class="StyledIconBase-ea9ulj-0 lbJwfL"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16.293 9.293 12 13.586 7.707 9.293l-1.414 1.414L12 16.414l5.707-5.707z"
          />
        </svg>
      </div>
    </div>
    <div
      class="h-0 overflow-hidden"
    >
      <div
        class="py-5"
      >
        <span
          class="mr-2"
        >
          17:18
        </span>
        <span
          class="mr-2"
        >
          25.6.2022
        </span>
        <span>
          Saturday
        </span>
      </div>
      <div
        class="first-letter: mt-4 px-1 flex"
      >
        <div
          class="first-letter: border border-solid border-primary w-8 h-8 text-lg rounded-[100%] flex items-center justify-center -mx-1"
        >
          <span
            class="transform -translate-y-[3%]"
          >
            w
          </span>
        </div>
        <div
          class="first-letter: border border-solid border-primary w-8 h-8 text-lg rounded-[100%] flex items-center justify-center -mx-1"
        >
          <span
            class="transform -translate-y-[3%]"
          >
            t
          </span>
        </div>
        <div
          class="first-letter: border border-solid border-primary w-8 h-8 text-lg rounded-[100%] flex items-center justify-center -mx-1"
        >
          <span
            class="transform -translate-y-[3%]"
          >
            x
          </span>
        </div>
      </div>
      <button
        class="btn btn-outline btn-accent mt-4 w-full"
        type="button"
      >
        Edit
      </button>
    </div>
  </div>
</div>
`;
